central/reports/scheduler/v2/schedule.go

```diff
@@ -0,0 +1,530 @@
+}
+
+func newSchedulerImpl(reportConfigDatastore reportConfigDS.DataStore, reportMetadataStore reportMetadataDS.DataStore,
+	reportSnapshotStore reportSnapshotDS.DataStore, notifierDatastore notifierDS.DataStore,
+	deploymentDatastore deploymentDS.DataStore, watchedImageDatastore watchedImageDS.DataStore,
+	collectionDatastore collectionDS.DataStore, collectionQueryRes collectionDS.QueryResolver,
+	notificationProcessor notifier.Processor, cronScheduler *cron.Cron, schema *graphql.Schema) *scheduler {
+	s := &scheduler{
+		reportConfigToEntryIDs:  make(map[string]cron.EntryID),
+		reportConfigDatastore:   reportConfigDatastore,
+		reportMetadataStore:     reportMetadataStore,
+		reportSnapshotStore:     reportSnapshotStore,
+		notifierDatastore:       notifierDatastore,
+		deploymentDatastore:     deploymentDatastore,
+		watchedImageDatastore:   watchedImageDatastore,
+		collectionDatastore:     collectionDatastore,
+		collectionQueryResolver: collectionQueryRes,
+		notificationProcessor:   notificationProcessor,
+		reportsQueue:            make([]*ReportRequest, 0),
+		sendNewReports:          make(chan bool, 200),
+		runningReportConfigIDs:  set.NewStringSet(),
+		Schema:                  schema,
+
+		stopper:         concurrency.NewStopper(),
+		cron:            cronScheduler,
+		concurrencySema: semaphore.NewWeighted(int64(env.ReportExecutionMaxConcurrency.IntegerSetting())),
+	}
+	return s
+}
+
+/* Concurrency and scheduling functions */
+
+func (s *scheduler) Start() {
+	go s.runReports()
+}
+
+func (s *scheduler) Stop() {
+	s.stopper.Client().Stop()
+	err := s.stopper.Client().Stopped().Wait()
+	if err != nil {
+		log.Errorf("Error stopping vulnerability report scheduler : %v", err)
+	}
+}
+
+func (s *scheduler) runReports() {
+	defer s.stopper.Flow().ReportStopped()
+	for {
+		select {
+		case <-s.stopper.Flow().StopRequested():
+			return
+		case <-s.sendNewReports:
+			reportRequest := s.selectNextRunnableReport()
+			if reportRequest == nil {
+				continue
+			}
+			if err := s.concurrencySema.Acquire(scheduledCtx, 1); err != nil {
+				log.Errorf("Error acquiring semaphore to run new report: %v", err)
+				continue
+			}
+			log.Infof("Executing report '%s' at %v", reportRequest.ReportConfig.GetName(), time.Now().Format(time.RFC822))
+			go s.runSingleReport(reportRequest)
+		}
+	}
+}
+
+func (s *scheduler) selectNextRunnableReport() *ReportRequest {
+	s.schedulerLock.Lock()
+	defer s.schedulerLock.Unlock()
+
+	runnableReportIdx := -1
+	for i, reportReq := range s.reportsQueue {
+		if !s.runningReportConfigIDs.Contains(reportReq.ReportConfig.GetId()) {
+			runnableReportIdx = i
+			break
+		}
+	}
+	if runnableReportIdx < 0 {
+		// did not find any reports to run
+		return nil
+	}
+	request := s.reportsQueue[runnableReportIdx]
+	s.reportsQueue = slices.Delete(s.reportsQueue, runnableReportIdx, runnableReportIdx+1)
```

Comment:
So in summary, we pull it off the queue of reports TO BE executed and put it on the queue of reports that ARE executing.  I know the variables kind of tell the tale, but I think some comments would help that either here or where the vars are defined.  As such, it may be less confusing to call `runningReportConfigIDs` simply `runningReports` perhaps.

Reply:
Yeah I can add comments where vars are declared in the scheduler struct. I will rename `runningReportConfigIDs` to `runningReportConfigs` , the reason to keep `config` in the name is to keep the distinction between a reportConfig and a report running for that config.